src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs

```diff
@@ -301,24 +301,29 @@ public string Name
         /// </remarks>
         public int GetNodeCount()
         {
+            // This method uses checked arithmetic to avoid silent overflows.
+            // It's impossible to parse a TypeName with NodeCount > int.MaxValue
+            // (TypeNameParseOptions.MaxNodes is an int), but it's possible
+            // to create such names with the Make* APIs.
             int result = 1;
 
             if (IsNested)
             {
-                result += DeclaringType.GetNodeCount();
+                result = checked(result + DeclaringType.GetNodeCount());
             }
-            else if (IsConstructedGenericType)
+
+            if (IsArray || IsPointer || IsByRef)
             {
-                result++;
+                result = checked(result + GetElementType().GetNodeCount());
             }
-            else if (IsArray || IsPointer || IsByRef)
+            else if (IsConstructedGenericType)
             {
-                result += GetElementType().GetNodeCount();
-            }
+                checked { result++; } // the generic type definition
```

Comment:
```suggestion
                result = checked(result + GetGenericTypeDefinition().GetNodeCount());
```
I think the logic would be easier to understand if it is written like this, and the `if (IsNested)` check is moved to `else if` after the `IsConstructedGenericType` check.